core/modules/system/system.module

```diff
@@ -4475,3 +4481,362 @@ function system_theme_debug_enabled_warning() {
     backdrop_set_message(t('Theme debugging is enabled. See the <a href="!url">Status report</a> page for more info.', array('!url' => url('admin/reports/status'))), 'warning', FALSE);
   }
 }
+
+/**
+ * Implements hook_process_html().
+ */
+function system_process_html(&$variables) {
+  backdrop_emit_cache_tags();
+  backdrop_emit_cache_max_age();
+}
+
+/**
+ * Implements hook_ajax_render_alter().
+ */
+function system_ajax_render_alter(&$commands) {
+  backdrop_emit_cache_tags();
+  backdrop_emit_cache_max_age();
+}
+
+/**
+ * Implements hook_emit_cache_tags().
+ */
+function system_emit_cache_tags($tags) {
+  $page_cache_tags = &backdrop_static(__FUNCTION__, array());
+
+  if (config_get('system.core.settings', 'emit_cache_tags')) {
+    backdrop_add_http_header('X-Backdrop-Cache-Tags', implode(' ', $tags));
+  }
+
+  // Store the emitted cache tags for further use.
+  $page_cache_tags = $tags;
+}
+
+/**
+ * Implements hook_emit_cache_max_age().
+ */
+function system_emit_cache_max_age($max_age) {
+  $page_cache_max_age = &backdrop_static(__FUNCTION__, array());
+
+  // Store the emitted cache max_age for further use.
+  $page_cache_max_age = $max_age;
+}
+
+/**
+ * Backwards compatible version of system_invalidate_cache_tags().
+ * This will automatically be used when the core patch to implement end of
```

Comment:
Is this needed? Since we are adding this to core, I figure the patch will be permanent, no?

Reply:
I have included the code for this and removed the related work around functions. https://www.drupal.org/project/drupal/issues/3004437